flatpak-proxy.c

```diff
@@ -837,13 +857,18 @@ buffer_read (ProxySide *side,
              Buffer    *buffer,
              GSocket   *socket)
 {
-  gsize received;
+  FlatpakProxyClient *client = side->client;
+  gsize received = 0;
   GInputVector v;
   GError *error = NULL;
   GSocketControlMessage **messages;
   int num_messages, i;
 
-  if (side->extra_input_data)
+  if (client->auth_state == AUTH_WAITING_FOR_BACKLOG &&
+      side == &client->client_side)
+    return FALSE;
+
+  if (side->extra_input_data && client->auth_state == AUTH_COMPLETE)
     {
       gsize extra_size;
       const guchar *extra_bytes = g_bytes_get_data (side->extra_input_data, &extra_size);
```

Comment:
A few lines later:
```c
side->extra_input_data =
            g_bytes_new_with_free_func
```
This overwrites `side->extra_input_data` without freeing the previous one.

Reply:
It frees the previous one by virtue of passing g_bytes_unref + side->extra_input_data as destroy notify.

Reply:
Wow, that's some extremely subtle code here... have never seen this kind of cleanup chain. What's the reason here instead of just cleaning up the old one immediately?

Reply:
Because the new GBytes references the memory of the first one. If we freed it early it would point into freed memory (or we would have to duplicate the memory).

Reply:
Its basically:
` extra_input_data = extra_input_data[received:]`

Reply:
I guess it would be cleaner if we always unref:ed it and explicitly ref:ed it in the re-use case.

Reply:
Well, it's working and this MR isn't touching it, so let's just resolve this thread.